### DISCUSSION TITLE
Please Help Me : to Install and Run the syson Project

### DISCUSSION THREAD

**Discussion by RootOfBigTree** [eclipse-syson/syson#745] [Q&A]
1. The README ( https://github.com/eclipse-syson/syson/blob/main/README.md ) has no installation and running.

2. And I do installation following the instructions of the doc ( https://doc.mbse-syson.org/syson/main/installation-guide/how-tos/install.html ) . However It's failed to do that ( both with Docker Compose and Manually ). The errors occur as following:

    (1) when doing the command: ` docker compose up`  after downloaded the [docker-compose](https://github.com/eclipse-syson/syson/blob/v2024.7.3/docker-compose.yml) file, the errors message :
         Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
         >>>> Executing external compose provider "/usr/local/bin/docker-compose". Please refer to the documentation for details. <<<<

         ERROR: yaml.scanner.ScannerError: mapping values are not allowed here
              in "./docker-compose.yml", line 170, column 46
         Error: executing /usr/local/bin/docker-compose up: exit status 1

    (2) when doing the command: `
                     docker run -p 5434:5432 --name syson-postgres \
                            -e POSTGRES_USER=dbuser \
                            -e POSTGRES_PASSWORD=dbpwd \
                            -e POSTGRES_DB=syson-db \
                            -d postgres
                     ` 
           the errors message :
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
✔ docker.io/library/postgres:latest
Trying to pull docker.io/library/postgres:latest...
WARN[0096] Failed, retrying in 1s ... (1/3). Error: copying system image from manifest list: parsing image configuration: Get "https://production.cloudflare.docker.com/registry-v2/docker/registry/v2/blobs/sha256/07/07a4ee949b9e7851d9ef860fd36f486aa821f23e1c17939244155fe825eaf79a/data?verify=1722955834-U6i8PDhTlPz4YTmjjYP5Ow%2FOxsQ%3D": dial tcp 128.121.243.76:443: i/o timeout 
WARN[0129] Failed, retrying in 1s ... (2/3). Error: copying system image from manifest list: parsing image configuration: Get "https://production.cloudflare.docker.com/registry-v2/docker/registry/v2/blobs/sha256/07/07a4ee949b9e7851d9ef860fd36f486aa821f23e1c17939244155fe825eaf79a/data?verify=1722955868-NcTxtIi2BGMXpQLUOVm1dT8isH8%3D": dial tcp 128.121.243.76:443: i/o timeout 
WARN[0164] Failed, retrying in 1s ... (3/3). Error: copying system image from manifest list: parsing image configuration: Get "https://production.cloudflare.docker.com/registry-v2/docker/registry/v2/blobs/sha256/07/07a4ee949b9e7851d9ef860fd36f486aa821f23e1c17939244155fe825eaf79a/data?verify=1722955902-w%2F5F451c9%2BOcJ7pWiqSBBjtxOpI%3D": dial tcp 128.121.243.76:443: i/o timeout 
Error: copying system image from manifest list: parsing image configuration: Get "https://production.cloudflare.docker.com/registry-v2/docker/registry/v2/blobs/sha256/07/07a4ee949b9e7851d9ef860fd36f486aa821f23e1c17939244155fe825eaf79a/data?verify=1722955937-ZILty0YeNkqb9Z%2BOE94WqL4h1Nc%3D": dial tcp 128.121.243.76:443: i/o timeout

3. As a result, it's for me that the docker-compose.yml file can NOT be dockered compose and the docker.io/library/postgres:latest can NOT be pulled.

4. So, I don't know how to install syson and run it following the instructions of the doc.

Please help me to tell me how to install and run the syson.

Thank you very much.

**Comment by AxelRICHARD** [comment]
Hello,

It seems you use podman instead of docker. We never tested installation through podman.
Please try with docker, and it should be ok.

Regards,

**Reply by mycr0ft** [reply]
The docker containers can be run via podman compose quite simply. The compose specification for OCI containers simply requires using the full address specification of repositories. Here is my working compose.yaml file: https://github.com/mycr0ft/SysON_podman_compose 

You should be able to use `podman compose up` from your user prompt if you have a reasonably modern podman install.

**Comment by RootOfBigTree** [comment]
Hi, AxelRICHARD:

1. I remove the podman: `dnf remove podman-docker.noarch` ;

2. Then install docker: `dnf install docker-ce docker-ce-cli containerd.io` ;

3. Then do: `docker compose up` 
    only occurs the message: 'Top-level object must be a mapping'
    but nothing.
 
I don't know the reason why can NOT docker compose up the docker-compose.yml file.

So, 

1. I install and configure the postgresql manually correctly, 

2. then run the syson application doing the command:
       `java -jar syson-application-2024.7.3.jar           --spring.datasource.url=jdbc:postgresql://localhost:5432/sysondb           --spring.datasource.username=sysonuser           --spring.datasource.password=sysonpasswd`

3. the end messages of ouput:
2024-08-07T15:32:51.545+08:00  INFO 10391 --- [           main] o.s.b.a.w.s.WelcomePageHandlerMapping    : Adding welcome page: class path resource [static/index.html]
2024-08-07T15:32:52.158+08:00  INFO 10391 --- [           main] o.s.b.a.e.web.EndpointLinksResolver      : Exposing 1 endpoint beneath base path '/actuator'
2024-08-07T15:32:52.298+08:00  INFO 10391 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port 8080 (http) with context path '/'
2024-08-07T15:32:52.313+08:00  INFO 10391 --- [           main] org.eclipse.syson.SysONApplication       : Started SysONApplication in 22.538 seconds (process running for 23.029)
2024-08-07T15:33:46.454+08:00  INFO 10391 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2024-08-07T15:33:46.455+08:00  INFO 10391 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2024-08-07T15:33:46.459+08:00  INFO 10391 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 4 ms

4. access the syson from url: localhost:8080/
    However, there is nothing in the browser page.

5. I open the DevTools of the Chrome browse, and it displays the message:
Uncaught TypeError: crypto.randomUUID is not a function
    at index-BamngbMA.js:201:8912
    at index-BamngbMA.js:1:24
    at index-BamngbMA.js:3911:4466

Please help me to tell me how to install and run the Syson .

Thank you so much.

**Comment by AxelRICHARD** [comment]
I'm sorry but I don't know why you have errors with docker on linux.
Some members on the team run SysON on linux and don't have any errors.

After a simple quick research, I found this: https://stackoverflow.com/questions/46550348/docker-stack-deploy-error-about-top-level-object-mappings

So I suggest to change double quotes for simple quotes in the docker-compose.yml file.

For the 5th problem, it seems you are accessing SysON from another machine than the one running the SysON application.
If you intend to install SysON on a local server or machine and allow several users to access it, you must enable HTTPS  (crypto.randomUUID is only available with HTTPS). Please read https://doc.mbse-syson.org/syson/main/installation-guide/how-tos/https.html for more details.

**Comment by RootOfBigTree** [comment]
Hi, AxelRICHARD,

1. I know the reason of the first problem of docker-compose.yml, that was cause by the download method that I do that from the command, `curl -O https://github.com/eclipse-syson/syson/blob/v2024.7.3/docker-compose.yml` . After downloaded, the file become a HTML type file. 

   And I copy the content of the file from page of https://github.com/eclipse-syson/syson/blob/v2024.7.3/docker-compose.yml to create a new docker-compose.yml . And do `docker compose up` . It looks OK. But maybe it's my network reason to cause timeout error 
`docker compose up`
WARN[0000] /home/zhuge/dev/tmp/syson/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion 
[+] Running 0/28
 ⠧ app [⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀] Pulling                                                                                                                                                                    159.9s 
   ⠴ 1671565cc8df Waiting                                                                                                                                                                        156.7s 
   ⠴ 3e94d13e55e7 Waiting                                                                                                                                                                        156.7s 
   ⠴ fa9c7528c685 Waiting                                                                                                                                                                        156.7s 
   ⠴ 53ad072f9cd1 Waiting                                                                                                                                                                        156.7s 
   ⠴ d6b983117533 Waiting                                                                                                                                                                        156.7s 
   ⠴ eac9917c3316 Waiting                                                                                                                                                                        156.7s 
   ⠴ ca44d2215c81 Waiting                                                                                                                                                                        156.7s 
   ⠴ 7ae8c505b227 Waiting                                                                                                                                                                        156.7s 
   ⠴ cea8eb14136a Waiting                                                                                                                                                                        156.7s 
   ⠴ 2c4d8bc67e39 Waiting                                                                                                                                                                        156.7s 
   ⠴ f496ace4b8db Waiting                                                                                                                                                                        156.7s 
   ⠴ a406a5043766 Waiting                                                                                                                                                                        156.7s 
 ⠇ database [⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀] Pulling                                                                                                                                                             159.9s 
   ⠦ efc2b5ad9eec Retrying in 1 second                                                                                                                                                           156.7s 
   ⠦ 1ba8510a018f Retrying in 1 second                                                                                                                                                           156.7s 
   ⠦ 31c1e1bb8144 Retrying in 1 second                                                                                                                                                           156.7s 
   ⠦ dde152d3f352 Waiting                                                                                                                                                                        156.7s 
   ⠦ 3b28f020f972 Waiting                                                                                                                                                                        156.7s 
   ⠦ bad5f36a70ca Waiting                                                                                                                                                                        156.7s 
   ⠦ f5484719b3aa Waiting                                                                                                                                                                        156.7s 
   ⠦ 2b7047a438a2 Waiting                                                                                                                                                                        156.7s 
   ⠦ 612a4adba252 Waiting                                                                                                                                                                        156.7s 
   ⠦ 81ecdb1b6073 Waiting                                                                                                                                                                        156.7s 
   ⠦ c3eff174fea9 Waiting                                                                                                                                                                        156.7s 
   ⠦ c916c46b125d Waiting                                                                                                                                                                        156.7s 
   ⠦ 4fe0272bc01a Waiting                                                                                                                                                                        156.7s 
   ⠦ 72ac7840aa1e Waiting                                                                                                                                                                        156.7s 
error pulling image configuration: download failed after attempts=6: dial tcp 199.96.63.177:443: i/o timeout

2. I enabled HTTPS according to  https://doc.mbse-syson.org/syson/main/installation-guide/how-tos/https.html  . And I can access SysON . 

3. But I do NOT know how to build and run SysON if I change something, because I don't find the building and running instructions in the README file. I don't find the SysON architecture too. Where are these documents?

Thank you very much.

**Comment by AxelRICHARD** [comment]
There is no build page yet in the documentation, no architecture document too.

We will try to update the documentation as soon as we have the time to do it.

But if you look at https://github.com/eclipse-syson/syson/blob/main/.github/workflows/build.yml from line 71 to line 87 you should be able to build SysON by yourself.

Regards,

**Comment by itafraze** [comment]
If it helps, I run SysON on rootless podman (startup code at https://gitlab.com/itafraze/podman-rootless-pod-syson). For development purposes, I guess you could replace the SysOn image with your own. Check [how it is built on GitHub](https://github.com/eclipse-syson/syson/blob/a531a9a40b28a8acc0daf33396ba2bbaa7d320bc/.github/workflows/build.yml) as starting point, as @AxelRICHARD said. Consider also #303
